wave_matching/src/icp.cpp

```diff
@@ -84,12 +130,152 @@ void ICPMatcher::estimateInfo() {
     switch (this->params.covar_estimator) {
         case ICPMatcherParams::covar_method::LUM: this->estimateLUM();
         case ICPMatcherParams::covar_method::CENSI: this->estimateCensi();
+        case ICPMatcherParams::covar_method::LUMold: this->estimateLUMold();
         default: return;
     }
 }
 
+void ICPMatcher::estimateLUMold() {
+    auto &source_trans = this->final;
+    PCLPointCloud *targetc;
+    if (this->params.res > 0) {
+        targetc = &(this->downsampled_target);
+    } else {
+        targetc = &(this->target);
+    }
+    int numSourcePts = source_trans->size();
+    std::vector<Eigen::Vector3f> corrs_aver(numSourcePts);
+    std::vector<Eigen::Vector3f> corrs_diff(numSourcePts);
+    int numCorr = 0;
+
+    wave::Mat6 edgeCov = wave::Mat6::Identity();
+
+    // build kd tree for source points
+    pcl::KdTreeFLANN<pcl::PointXYZ> kdtree;
+    if (this->params.res > 0) {
+    }
+    kdtree.setInputCloud(*targetc);
+
+    // iterate through the source cloud and compute match covariance
+
+    for (int i = 0; i < numSourcePts; i++) {
+        pcl::PointXYZ qpt = source_trans->points[i];
+        std::vector<int> nn_idx;
+        std::vector<float> nn_sqr_dist;
+        kdtree.nearestKSearch(
+          qpt,
+          1,
+          nn_idx,
+          nn_sqr_dist);  // returns the index of the nn point in the targetc
+
+        if (nn_sqr_dist[0] < this->params.max_corr *
+                               this->params.max_corr)  // if the distance to
+                                                       // point is less than max
+                                                       // correspondence
+                                                       // distance, use it to
+                                                       // calculate
+        {
+            Eigen::Vector3f source_pt = qpt.getVector3fMap();
+            Eigen::Vector3f target_pt =
+              (*targetc)->points[nn_idx[0]].getVector3fMap();
+
+            // Compute the point pair average and difference and store for later
+            // use
+            corrs_aver[numCorr] = 0.5 * (source_pt + target_pt);
+            corrs_diff[numCorr] = source_pt - target_pt;
+            numCorr++;
+        } else {
+            continue;
+        }
+    }
+    corrs_aver.resize(numCorr);
+    corrs_diff.resize(numCorr);
+
+    // now compute the M matrix
+    wave::Mat6 MM = wave::Mat6::Zero();
+    wave::Vec6 MZ = wave::Vec6::Zero();
+    for (int ci = 0; ci != numCorr; ++ci)  // ci = correspondence iterator
+    {
+        // Fast computation of summation elements of M'M
+        MM(0, 4) -= corrs_aver[ci](1);
```

Comment:
Hard to read.
- use range-based for (`for(const auto &element : vector)`)
- if possible write a mathematical expression using Eigen's helpful methods instead of manually indexing.

Reply:
This function is from PCL, I don't want to change it because it has presumably been tested to work